File: modules/agent/src/schemas/organization.py
"""Organization schemas for API validation."""

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, field_validator

from lib.validators import validate_phone


class OrganizationCreate(BaseModel):
    name: str
    website: Optional[str] = None
    phone: Optional[str] = None
    employee_count: Optional[int] = None
    employee_count_range_id: Optional[int] = None
    funding_stage_id: Optional[int] = None
    description: Optional[str] = None
    account_id: Optional[int] = None
    industry_ids: Optional[List[int]] = None
    project_ids: Optional[List[int]] = None
    revenue_range_id: Optional[int] = None
    founding_year: Optional[int] = None
    headquarters_city: Optional[str] = None
    headquarters_state: Optional[str] = None
    headquarters_country: Optional[str] = None
    company_type: Optional[str] = None
    linkedin_url: Optional[str] = None
    crunchbase_url: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)

    @field_validator("founding_year")
    @classmethod
    def validate_founding_year(cls, v):
        if v is None:
            return v
        current_year = datetime.now().year
        if v < 1800 or v > current_year:
            raise ValueError(f"founding_year must be between 1800 and {current_year}")
        return v


class OrganizationUpdate(BaseModel):
    name: Optional[str] = None
    website: Optional[str] = None
    phone: Optional[str] = None
    employee_count: Optional[int] = None
    employee_count_range_id: Optional[int] = None
    funding_stage_id: Optional[int] = None
    description: Optional[str] = None
    industry_ids: Optional[List[int]] = None
    project_ids: Optional[List[int]] = None
    revenue_range_id: Optional[int] = None
    founding_year: Optional[int] = None
    headquarters_city: Optional[str] = None
    headquarters_state: Optional[str] = None
    headquarters_country: Optional[str] = None
    company_type: Optional[str] = None
    linkedin_url: Optional[str] = None
    crunchbase_url: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)

    @field_validator("founding_year")
    @classmethod
    def validate_founding_year(cls, v):
        if v is None:
            return v
        current_year = datetime.now().year
        if v < 1800 or v > current_year:
            raise ValueError(f"founding_year must be between 1800 and {current_year}")
        return v


class OrgContactCreate(BaseModel):
    individual_id: Optional[int] = None
    first_name: str
    last_name: str
    title: Optional[str] = None
    department: Optional[str] = None
    role: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    is_primary: bool = False
    notes: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class OrgContactUpdate(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None
    title: Optional[str] = None
    department: Optional[str] = None
    role: Optional[str] = None
    email: Optional[str] = None
    phone: Optional[str] = None
    is_primary: Optional[bool] = None
    notes: Optional[str] = None

    @field_validator("phone")
    @classmethod
    def validate_phone_format(cls, v):
        return validate_phone(v)


class OrgTechnologyCreate(BaseModel):
    technology_id: int
    source: Optional[str] = None
    confidence: Optional[float] = None


class FundingRoundCreate(BaseModel):
    round_type: str
    amount: Optional[int] = None
    announced_date: Optional[str] = None
    lead_investor: Optional[str] = None
    source_url: Optional[str] = None


class SignalCreate(BaseModel):
    signal_type: str
    headline: str
    description: Optional[str] = None
    signal_date: Optional[str] = None
    source: Optional[str] = None
    source_url: Optional[str] = None
    sentiment: Optional[str] = None
    relevance_score: Optional[float] = None

File: modules/agent/internal/schemas/organization.go
package schemas

// OrganizationCreate represents the request body for creating an organization
type OrganizationCreate struct {
	Name                 string   `json:"name" validate:"required"`
	Website              *string  `json:"website" validate:"omitempty,url"`
	Phone                *string  `json:"phone" validate:"omitempty,e164"`
	EmployeeCount        *int     `json:"employee_count" validate:"omitempty,min=0"`
	EmployeeCountRangeID *int64   `json:"employee_count_range_id"`
	FundingStageID       *int64   `json:"funding_stage_id"`
	Description          *string  `json:"description"`
	AccountID            *int64   `json:"account_id"`
	IndustryIDs          []int64  `json:"industry_ids"`
	ProjectIDs           []int64  `json:"project_ids"`
	RevenueRangeID       *int64   `json:"revenue_range_id"`
	FoundingYear         *int     `json:"founding_year" validate:"omitempty,min=1800,max=2100"`
	HeadquartersCity     *string  `json:"headquarters_city"`
	HeadquartersState    *string  `json:"headquarters_state"`
	HeadquartersCountry  *string  `json:"headquarters_country"`
	CompanyType          *string  `json:"company_type" validate:"omitempty,oneof=private public nonprofit government"`
	LinkedInURL          *string  `json:"linkedin_url" validate:"omitempty,url"`
	CrunchbaseURL        *string  `json:"crunchbase_url" validate:"omitempty,url"`
}

// OrganizationUpdate represents the request body for updating an organization
type OrganizationUpdate struct {
	Name                 *string  `json:"name"`
	Website              *string  `json:"website" validate:"omitempty,url"`
	Phone                *string  `json:"phone" validate:"omitempty,e164"`
	EmployeeCount        *int     `json:"employee_count" validate:"omitempty,min=0"`
	EmployeeCountRangeID *int64   `json:"employee_count_range_id"`
	FundingStageID       *int64   `json:"funding_stage_id"`
	Description          *string  `json:"description"`
	IndustryIDs          []int64  `json:"industry_ids"`
	ProjectIDs           []int64  `json:"project_ids"`
	RevenueRangeID       *int64   `json:"revenue_range_id"`
	FoundingYear         *int     `json:"founding_year" validate:"omitempty,min=1800,max=2100"`
	HeadquartersCity     *string  `json:"headquarters_city"`
	HeadquartersState    *string  `json:"headquarters_state"`
	HeadquartersCountry  *string  `json:"headquarters_country"`
	CompanyType          *string  `json:"company_type" validate:"omitempty,oneof=private public nonprofit government"`
	LinkedInURL          *string  `json:"linkedin_url" validate:"omitempty,url"`
	CrunchbaseURL        *string  `json:"crunchbase_url" validate:"omitempty,url"`
}

// OrgContactCreate represents the request body for adding a contact to an organization
type OrgContactCreate struct {
	IndividualID *int64  `json:"individual_id"`
	FirstName    string  `json:"first_name" validate:"required"`
	LastName     string  `json:"last_name" validate:"required"`
	Title        *string `json:"title"`
	Department   *string `json:"department"`
	Role         *string `json:"role"`
	Email        *string `json:"email" validate:"omitempty,email"`
	Phone        *string `json:"phone" validate:"omitempty,e164"`
	IsPrimary    bool    `json:"is_primary"`
	Notes        *string `json:"notes"`
}

// OrgContactUpdate represents the request body for updating an organization contact
type OrgContactUpdate struct {
	FirstName  *string `json:"first_name"`
	LastName   *string `json:"last_name"`
	Title      *string `json:"title"`
	Department *string `json:"department"`
	Role       *string `json:"role"`
	Email      *string `json:"email" validate:"omitempty,email"`
	Phone      *string `json:"phone" validate:"omitempty,e164"`
	IsPrimary  *bool   `json:"is_primary"`
	Notes      *string `json:"notes"`
}

// OrgTechnologyCreate represents the request body for adding technology to an organization
type OrgTechnologyCreate struct {
	TechnologyID int64    `json:"technology_id" validate:"required"`
	Source       *string  `json:"source"`
	Confidence   *float64 `json:"confidence" validate:"omitempty,min=0,max=1"`
}

// FundingRoundCreate represents the request body for creating a funding round
type FundingRoundCreate struct {
	RoundType     string  `json:"round_type" validate:"required"`
	Amount        *int64  `json:"amount" validate:"omitempty,min=0"`
	AnnouncedDate *string `json:"announced_date"`
	LeadInvestor  *string `json:"lead_investor"`
	SourceURL     *string `json:"source_url" validate:"omitempty,url"`
}

// SignalCreate represents the request body for creating a company signal
type SignalCreate struct {
	SignalType     string   `json:"signal_type" validate:"required"`
	Headline       string   `json:"headline" validate:"required"`
	Description    *string  `json:"description"`
	SignalDate     *string  `json:"signal_date"`
	Source         *string  `json:"source"`
	SourceURL      *string  `json:"source_url" validate:"omitempty,url"`
	Sentiment      *string  `json:"sentiment" validate:"omitempty,oneof=positive negative neutral"`
	RelevanceScore *float64 `json:"relevance_score" validate:"omitempty,min=0,max=1"`
}

// OrgRelationshipCreate represents the request body for creating an organization relationship
type OrgRelationshipCreate struct {
	ToOrganizationID int64   `json:"to_organization_id" validate:"required"`
	RelationshipType *string `json:"relationship_type"`
	Notes            *string `json:"notes"`
}
